phpBB/phpbb/ucp/controller/webpush.php

```diff
@@ -0,0 +1,241 @@
+			'U_WEBPUSH_GET_NOTIFICATION'	=> $this->controller_helper->route('phpbb_ucp_push_get_notification_controller'),
+		]);
+
+		$response = new Response($content);
+		$response->headers->set('Content-Type', 'text/javascript; charset=UTF-8');
+
+		if (!empty($this->user->data['is_bot']))
+		{
+			// Let reverse proxies know we detected a bot.
+			$response->headers->set('X-PHPBB-IS-BOT', 'yes');
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Get template variables for subscribe type pages
+	 *
+	 * @return array
+	 */
+	protected function get_subscribe_vars(): array
+	{
+		return [
+			'U_WEBPUSH_SUBSCRIBE'	=> $this->controller_helper->route('phpbb_ucp_push_subscribe_controller'),
+			'U_WEBPUSH_UNSUBSCRIBE'	=> $this->controller_helper->route('phpbb_ucp_push_unsubscribe_controller'),
+			'FORM_TOKENS'			=> $this->form_helper->get_form_tokens(self::FORM_TOKEN_UCP),
+		];
+	}
+
+	/**
+	 * Check (un)subscribe form for valid link hash
+	 *
+	 * @throws http_exception If form is invalid or user should not request (un)subscription
+	 * @return void
+	 */
+	protected function check_subscribe_requests(): void
+	{
+		if (!$this->form_helper->check_form_tokens(self::FORM_TOKEN_UCP))
+		{
+			throw new http_exception(Response::HTTP_BAD_REQUEST, 'FORM_INVALID');
+		}
+
+		// Subscribe should only be available for logged-in "normal" users
+		if (!$this->request->is_ajax() || $this->user->id() == ANONYMOUS || $this->user->data['is_bot']
+			|| $this->user->data['user_type'] == USER_IGNORE || $this->user->data['user_type'] == USER_INACTIVE)
+		{
+			throw new http_exception(Response::HTTP_FORBIDDEN, 'NO_AUTH_OPERATION');
+		}
+	}
+
+	/**
+	 * Handle subscribe requests
+	 *
+	 * @param symfony_request $symfony_request
+	 * @return JsonResponse
+	 */
+	public function subscribe(symfony_request $symfony_request): JsonResponse
+	{
+		$this->check_subscribe_requests();
+
+		$data = json_sanitizer::decode($symfony_request->get('data', ''));
+
+		$sql = 'INSERT INTO ' . $this->push_subscriptions_table . ' ' . $this->db->sql_build_array('INSERT', [
+			'user_id'			=> $this->user->id(),
+			'endpoint'			=> $data['endpoint'],
+			'expiration_time'	=> $data['expiration_time'] ?? 0,
+			'p256dh'			=> $data['keys']['p256dh'],
+			'auth'				=> $data['keys']['auth'],
+		]);
+		$this->db->sql_query($sql);
+
+		return new JsonResponse([
+			'success'		=> true,
+			'form_tokens'	=> $this->form_helper->get_form_tokens(self::FORM_TOKEN_UCP),
+		]);
+	}
+
+	/**
+	 * Handle unsubscribe requests
+	 *
+	 * @param symfony_request $symfony_request
+	 * @return JsonResponse
+	 */
+	public function unsubscribe(symfony_request $symfony_request): JsonResponse
+	{
+		$this->check_subscribe_requests();
+
+		$data = json_sanitizer::decode($symfony_request->get('data', ''));
+
+		$endpoint = $data['endpoint'];
+
+		$sql = 'DELETE FROM ' . $this->push_subscriptions_table . '
+			WHERE user_id = ' . (int) $this->user->id() . "
+				AND endpoint = '" . (int) $this->db->sql_escape($endpoint) . "'";
```

Comment:
You're sql_escaping an int-casted variable here. Is `$endpoint` a string?

Reply:
Endpoint is a string and hence the `(int)` shouldn't be there :)